packages/material-ui/src/ListItem/ListItem.js

```diff
@@ -179,9 +182,35 @@ ListItem.propTypes = {
    */
   button: PropTypes.bool,
   /**
-   * The content of the component.
+   * The content of the component. If a `ListItemSecondaryAction` is used it must
+   * be the last child.
    */
-  children: PropTypes.node,
+  children: chainPropTypes(PropTypes.node, props => {
+    const children = React.Children.toArray(props.children);
+
+    // React.Children.toArray(props.children).findLastIndex(isListItemSecondaryAction)
```

Comment:
What about `.findIndex()` and replace it once we drop IE 11?

Reply:
I was just asking about the commented out code :) The below code looks okay to me

Reply:
Yeah. I agree, I would leave something about IE 11 so we don't forget to refactor the code.

Reply:
comment explains the intent of the code

Reply:
Won't using findIndex be better than findLastIndex? The current logic won't detect the usage of two ListItemSecondaryAction.

Reply:
two secondary actions were always ok. It was only problematic if none of the secondary actions were the last child. This PR only adds a warning for existing behavior. If you want a warning for multiple secondary actions then we can include that too.

Reply:
We're getting of track again what the scope of this PR was about. It was aimed at adressing the points made in #13916. Nobody ever mentioned that we should warn against multiple secondary actions used. If you want a warning for that then sure we can add it. Again these are two different issues and if you want to combine those I have nothing against that. Just need to make sure we understand that.

Reply:
My point is: we get the multiple secondary warning for free if we reverse the logic from findIndex to findLastIndex.

Reply:
No we simply display the wrong warning in it's current state if people use `<ListItem><Action /><Text /><Action /></ListItem>`. I would really appreciate it if we could focus on the original intent and not derail this again to do more than it's supposed to without explicitly talking about it. It's handwaving and pretty draining personally.

It's not low hanging fruit if you ask for an additional warning.

Reply:
I'm sorry, I don't understand. The warning is:

> Material-UI: you used an element after ListItemSecondaryAction.
For ListItem to detect that it has a secondary action you must pass it has the last children to ListItem.

It can very well warn for 
```jsx
<ListItem><Action /><Text /><Action /></ListItem>
```

Reply:
From the PR perspective, it means doing:
```diff
-// React.Children.toArray(props.children).findLastIndex(isListItemSecondaryAction)
-for (let i = children.length - 1; i >= 0; i -= 1) {
+// To replace with findIndex once we stop IE 11 support
+for (let i = 0; i < children.length; i += 1) {
```

Reply:
> It's handwaving and pretty draining personally.

Yes, I understand it. I admire your motivation for adding a new warning. Now, know that we will likely not come back to the issue before a long time given how many important issues we have to handle. It's why I wanted to see if, with a bit more effort, we can do better. We might.

Reply:
`<ListItem><Action /><Text /><Action /></ListItem>` is invalid. The existing warning is good enough: the action should be the last child. Wtf it's already the last child. Oh but I have another action, OK it should be the last child too. Wait, I can only have one child. Ok so I can only use one action.

Reply:
I'm wondering if React don't already warn about button > button. It would just add another hint.

Reply:
Anyway, it's up to you. It's not truly important. We should invest too much time here.

Reply:
> Wait, I can only have one child

You're making my argument here. The current warning with your implementation proposal would not warn about "multiple children detected". It would create exactly the scenario you described.

We never warned explicitly about multiple actions. I don't know of any issue that described confusion about multiple actions. Maybe people are already using multiple actions just fine. If you want to warn against multiple actions feel free to open a separate PR.